delivery_auto_refresh/models/sale_order.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2018 Tecnativa - Pedro M. Baeza
+# Copyright 2021 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+from odoo.tools import safe_eval
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    carrier_id = fields.Many2one(
+        compute="_compute_carrier_id", store=True, readonly=False
+    )
+
+    @api.depends("partner_id")
+    def _compute_carrier_id(self):
+        if hasattr(super(), "_compute_carrier_id"):
+            super()._compute_carrier_id()
+        for rec in self:
+            if rec.partner_id:
+                if rec.partner_id.property_delivery_carrier_id:
+                    rec.carrier_id = rec.partner_id.property_delivery_carrier_id
+                else:
+                    rec.carrier_id = self.env["delivery.carrier"].search(
+                        [
+                            "|",
+                            ("company_id", "=", False),
+                            ("company_id", "=", self.company_id.id),
+                        ],
+                        limit=1,
+                    )
+
+    def _get_param_auto_add_delivery_line(self):
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "delivery_auto_refresh.auto_add_delivery_line"
+        return safe_eval(get_param(param, "0"))
+
+    def _auto_refresh_delivery(self):
+        self.ensure_one()
+        # Make sure that if you have removed the carrier, the line is gone
+        if self.state in {"draft", "sent"}:
+            # Context added to avoid the recursive calls and save the new
+            # value of carrier_id
+            ctx = self.env.context.copy()
+            ctx["new_carrier_id"] = self.carrier_id.id
+            self.env.context = ctx
```

Comment:
This can't be done this way, as you are tricking the system through Python internals.